pkg/relayer/session/sessiontree.go

```diff
@@ -19,7 +19,11 @@ import (
 var _ relayer.SessionTree = (*sessionTree)(nil)
 
 // sessionTree is an implementation of the SessionTree interface.
-// TODO_TEST: Add tests to the sessionTree.
+// TODO_BETA(@red-0ne): Per the Relay Mining paper, we need to optimistically store
+// the number of requests that an application can pay for. This needs to be tracked
+// based on the app's stake in the beginning of a session and the number of nodes
+// per session. An operator should be able to specify "overservicing_okay" whereby
```

Comment:
I'm not sure keeping track of the app's stake would be relevant.

This is because at the beginning of a new session, there may be one (or more) settlements from previous sessions that will occur after the new session starts (due to C&P window offsets) and that will alter the stake.

Reply:
I'm changing this to a `TODO_IN_THIS_PR_DISCUSS` until we are in alignment.

@red-0ne This is the foundation of optimistic rate limiting and the best we can do. The app's stake in escrow is a signal into rate limiting. 

Window-base centralized rate limiting solutions are better (because it's 100% centralized and controlled).

We need to be in alignment that this is the goal, or we likely have to go back to the drawing board prior to mainNet.

Reply:
@Olshansk I agree and think that max claimable amount with stake check is the best coordination-free solution to over servicing.

My only concern is that the relay miners will rely on an outdated stake (at least by one update). I spent quite some time trying to workaround this but no solution not involving off-chain suppliers cooperation came to mind.

The best mitigation I can imagine is suppliers assuming that the application consumed max claimable amount for each non-yet-settled closed session (probably 1 or 2 sessions).

To conclude, I think we are aligned that we'll have to use max claimable amount and app stake in a more or less elaborated way to rate-limit the application. Future solutions will be added when appropriate.